pkg/controller/serviceaccount/legacy_serviceaccount_token_cleaner.go

```diff
@@ -0,0 +1,274 @@
+		legacySATokenCleanUpInterval: options.LegacySATokenCleanUpInterval,
+	}
+
+	return tc, nil
+}
+
+func (tc *LegacySATokenCleaner) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+
+	logger := klog.FromContext(ctx)
+	logger.Info("Starting legacy service account token cleaner controller")
+	defer logger.Info("Shutting down legacy service account token cleaner controller")
+
+	if !cache.WaitForNamedCacheSync("legacy-service-account-token-cleaner", ctx.Done(), tc.saInformerSynced, tc.secretInformerSynced, tc.configMapInformerSynced, tc.podInformerSynced) {
+		return
+	}
+
+	go wait.UntilWithContext(ctx, tc.evaluateSATokens, tc.legacySATokenCleanUpInterval)
+
+	<-ctx.Done()
+}
+
+func (tc *LegacySATokenCleaner) evaluateSATokens(ctx context.Context) {
+	logger := klog.FromContext(ctx)
+	secretList, err := tc.secretLister.Secrets(metav1.NamespaceAll).List(labels.Everything())
+	if err != nil {
+		logger.Error(err, "Getting secret list from secret cache")
+		return
+	}
+	for _, secret := range secretList {
+		if secret.Type != v1.SecretTypeServiceAccountToken {
+			continue
+		}
+		sa, saErr := tc.getServiceAccount(secret)
+		switch {
+		case saErr != nil:
+			logger.Error(saErr, "Getting service account")
+		case sa == nil:
+			logger.V(4).Info("Service account does not exist", "secret", klog.KRef(secret.Namespace, secret.Name))
+		case tc.shouldCleanToken(logger, secret, sa):
+			logger.Info("Auto-generated service account token is not used for more than 365 * 24 hours, deleting token", "secret", klog.KRef(secret.Namespace, secret.Name))
+			if err := tc.client.CoreV1().Secrets(secret.Namespace).Delete(ctx, secret.Name, metav1.DeleteOptions{}); err != nil {
+				logger.Error(err, "Deleting legacy service account token", "secret", klog.KRef(secret.Namespace, secret.Name), "serviceaccount", sa.Name)
+			}
+		}
+	}
+}
+
+func (tc *LegacySATokenCleaner) getServiceAccount(secret *v1.Secret) (*v1.ServiceAccount, error) {
+	saName := secret.Annotations[v1.ServiceAccountNameKey]
+	saUID := types.UID(secret.Annotations[v1.ServiceAccountUIDKey])
+	sa, err := tc.saLister.ServiceAccounts(secret.Namespace).Get(saName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	// Ensure UID matches if given
+	if len(saUID) == 0 || saUID == sa.UID {
+		return sa, nil
+	}
+
+	return nil, nil
+}
+
+func (tc *LegacySATokenCleaner) shouldCleanToken(logger klog.Logger, secret *v1.Secret, sa *v1.ServiceAccount) bool {
+	// determine whether the secret is referenced by the service account, if not referenced, the secret is manually generated and should not be cleaned up
+	if !hasSecretReference(sa, secret.Name) {
+		return false
+	}
+
+	// determine whether the period since last usage has exceeded the legacySATokenCleanUpPeriod
+	if !tc.exceedLegacySATokenCleanUpPeriod(logger, secret) {
+		return false
+	}
+
+	return !tc.isPodMounted(logger, secret)
```

Comment:
this check can be relatively expensive for namespaces with a lot of pods... if we can avoid computing this multiple times for the same namespace in the same sync that would be ideal. I made a suggestion above about changing this to `getMountedSecretNames` and doing that at most once per namespace in `evaluateSATokens`